src/environment/renderer.go

```diff
@@ -0,0 +1,281 @@
+	// TODO
+	// This will probably only need to be done once, so maybe do it after parsing (not before rendering)?
+	return nil
+}
+
+func (r *renderer) renderTemplate(node *nodes.Template, frame *frame) (iterator.Iterator[string], error) {
+	return iterator.Map(iterator.FromSlice(node.Body), func(n nodes.Node) (string, error) {
+		if err := r.renderNode(n, frame); err != nil {
+			return "", err
+		}
+		s := r.out.builder.String()
+		r.out.builder.Reset()
+		return s, nil
+	}), nil
+}
+
+func (r *renderer) renderNode(node nodes.Node, frame *frame) error {
+	switch n := node.(type) {
+	case *nodes.Extends:
+		panic("not implemented")
+	case *nodes.Macro:
+		panic("not implemented")
+	case *nodes.Scope:
+		panic("not implemented")
+	case *nodes.FilterBlock:
+		panic("not implemented")
+	case *nodes.Output:
+		return r.renderOutput(n, frame)
+	case *nodes.If:
+		return r.renderIf(n, frame)
+	case *nodes.ScopedEvalContextModifier:
+		panic("not implemented")
+	case *nodes.EvalContextModifier:
+		panic("not implemented")
+	case *nodes.CallBlock:
+		panic("not implemented")
+	case *nodes.Assign:
+		panic("not implemented")
+	case *nodes.AssignBlock:
+		panic("not implemented")
+	case *nodes.With:
+		panic("not implemented")
+	case *nodes.For:
+		panic("not implemented")
+	case *nodes.Block:
+		panic("not implemented")
+	// Include, Import, FromImport will probably require different handling (in renderTemplate)
+	case *nodes.Include:
+		panic("not implemented")
+	case *nodes.Import:
+		panic("not implemented")
+	case *nodes.FromImport:
+		panic("not implemented")
+	default:
+		// TODO it's possible that some of the nodes are missing in the switch, check this in the future
+		panic(fmt.Sprintf("unexpected node type `%v` (this is a bug in gojinja)", node))
+	}
+}
+
+func (r *renderer) blockvisit(body []nodes.Node, frame *frame) error {
+	for _, n := range body {
+		if err := r.renderNode(n, frame); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *renderer) renderIf(node *nodes.If, frame *frame) error {
+	ifFrame := frame.soft()
+
+	cond, err := coerceBool(r.ctx, frame, r, node.Test)
+	if err != nil {
+		return err
+	}
+
+	// if body
+	if cond {
+		return r.blockvisit(node.Body, ifFrame)
+	}
+	// else-ifs
+	for _, elif := range node.Elif {
+		cond, err := coerceBool(r.ctx, frame, r, node.Test)
+		if err != nil {
+			return err
+		}
+		if cond {
+			return r.blockvisit(elif.Body, ifFrame)
+		}
+	}
+	// else
+	return r.blockvisit(node.Else, ifFrame)
+}
+
+func (r *renderer) renderOutput(node *nodes.Output, frame *frame) error {
+	if frame.requireOutputCheck {
+		if r.hasKnownExtends {
+			return nil
+		}
+		// if !parentTemplate != nil {  // TODO
+		// 	  return nil
+		// }
+	}
+
+	for _, expr := range node.Nodes {
+		if err := r.writeOutputChild(expr, frame); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func identity[T any](x T) T {
+	return x
+}
+
+func (r *renderer) write(v any) error {
```

Comment:
imho this should take `string`

Reply:
I think it has to take any (required for native move support, jinja has a bit of special handling for it)